src/info/repo.rs

```diff
@@ -16,159 +15,212 @@ use time::OffsetDateTime;
 use time_humanize::HumanTime;
 
 pub struct Repo<'a> {
-    repo: &'a Repository,
-    logs: Vec<Commit<'a>>,
+    git2_repo: &'a Repository,
+    repo: git::Repository,
+    authors: Vec<Author>,
+    total_num_authors: usize,
+    num_commits: usize,
+    /// false if we have found the first commit that started it all, true if the repository is shallow.
+    is_shallow: bool,
+    time_of_most_recent_commit: git::actor::Time,
+    time_of_first_commit: git::actor::Time,
 }
 
-#[derive(Hash, PartialEq, Eq)]
 pub struct Sig {
-    name: String,
-    email: String,
+    name: git::bstr::BString,
+    email: git::bstr::BString,
+    email_lowercase: Option<String>,
```

Comment:
From Git mailmap [documentation](https://git-scm.com/docs/gitmailmap):

>Both E-Mails and names are matched case-insensitively.

In other words, Git is case-insensitive **only** when looking up entries in the `.mailmap` file, otherwise it is case-sensitive.

As a consequence, the email_lowercase property of `Sig` should not exist and it should be the responsibility of `gitoxide`'s mailmap module to be case insensitive.

Reply:
Even though `git-mailmap` does exactly that, it only does it for emails that are actually in the `.mailmap`. There are a few cases of contributors with `Foo Bar <foo@example.com` and `Foo Bar <FoO@Example.com>` email addresses in the linux kernel repo for example which need this code to be deduplicated as inarguably they are the same contributor (e.g. sending an email to `FoO@Example.com` and `foo@example.com` both reach `Foo Bar`).

Thus I do find this behaviour more in line with what the contributor count is trying to accomplish, but see why one would want to stick to what git does even if it _potentially_ is less correct.

The additional case-deduplication in `Sig` was removed in fb4d449134eff556aac1d14fb95ee9243b0e747e.
The linux kernel v5.16 now counts 31705 contributors, `git shortlog -sne | wc -l` counts 31700, and mainline onefetch counts 31711.

I looked into why this is a difference but gave up when I realized that for one it's hard to diff the output as git only sorts by number, providing a different sorting for those who have the same commit count which `onefetch` now addresses by also sorting by name. And then I saw the last entries in shortlog, with the last one being `<>` and thought that there probably is no reason to strive for being more like it.

<details><summary>shortlogs by git and onefetch</summary>
<br>
`git shortlog -sne` on `linux@v5.16` (yes, the output of unicode characters is somehow processed or unreadable)

```
[..] last 29 lines
     1	zoro <long17.cool@163.com>
     1	Ãrjan Eide <orjan.eide@arm.com>
     1	Ãvar ArnfjÃ¶rÃ° Bjarmason <avarab@gmail.com>
     1	Ãtienne Bersac <bersace@gmail.com>
     1	Ãmer Sezgin Ugurlu <omer.ugurlu@a-kent.com>
     1	Åukasz Bartosik <lb@semihalf.com>
     1	Åukasz Daniluk <lukasz.daniluk@intel.com>
     1	ÐÐ»ÐµÐºÑÐµÐ¹ ÐÑÐ°Ð¼Ð°ÑÐµÐ½ÐºÐ¾ <alexeyk13@yandex.ru>
     1	ÐÑÐ´Ð¸ Ð Ð¾Ð¼Ð°Ð½ÑÐ¾, AreMa Inc <knightrider@are.ma>
     1	ÐÐ¸ÐºÐ¾Ð»Ð°Ð¹ ÐÑÐ´ÑÑÐ²ÑÐµÐ² <nkudriavtsev@gmail.com>
     1	Ð ÑÑÐ»Ð°Ð½ ÐÑÐ°ÐµÐ² <ubijca16@gmail.com>
     1	âCosmin <cosmin90stanescu@gmail.com>
     1	ã»ã¡ <knightrider@are.ma>
     1	å¤·å(Caspar) <jinli.zjl@alibaba-inc.com>
     1	å®éæ¶ <songjinshi@xiaomi.com>
     1	å¼ å <zjn77777@163.com>
     1	å¼ å¿ å±± <zzs0213@gmail.com>
     1	å½­ä¸æ <pengdonglin@smartisan.com>
     1	å½­æµ©(Richard) <richard.peng@oppo.com>
     1	æ´ªä¸ç«¹ <sam.hung@emc.com.tw>
     1	çæ¯æ° <zhenjie.wang@sjtu.edu.cn>
     1	çæè <wenhu.wang@vivo.com>
     1	çç¨å <wangchenggang@vivo.com>
     1	è¯´ä¸å¾ <gavin.kx@qq.com>
     1	é©¬å»ºæ <majianpeng@gmail.com>
     1	é»ä¹ <huangle1@jd.com>
     1	ê¹ê·ì <chammoru@gmail.com>
     1	ë¨ìë¯¼ <youngmin.nam@samsung.com>
     1	ë°°ìì§ <soukjin.bae@samsung.com>
     1	ì´ê±´í¸ <urpapa@gmail.com>
     1	 <>
```

Debug output of `onefetch`.

```
[..] last 29 lines
1 zoro <long17.cool@163.com>
1 Ãrjan Eide <orjan.eide@arm.com>
1 Ævar Arnfjörð Bjarmason <avarab@gmail.com>
1 Étienne Bersac <bersace@gmail.com>
1 Ömer Sezgin Ugurlu <omer.ugurlu@a-kent.com>
1 Łukasz Bartosik <lb@semihalf.com>
1 Łukasz Daniluk <lukasz.daniluk@intel.com>
1 Алексей Крамаренко <alexeyk13@yandex.ru>
1 Буди Романто, AreMa Inc <knightrider@are.ma>
1 Николай Кудрявцев <nkudriavtsev@gmail.com>
1 Руслан Исаев <ubijca16@gmail.com>
1 “Cosmin <cosmin90stanescu@gmail.com>
1 ほち <knightrider@are.ma>
1 夷则(Caspar) <jinli.zjl@alibaba-inc.com>
1 宋金时 <songjinshi@xiaomi.com>
1 张君 <zjn77777@163.com>
1 张忠山 <zzs0213@gmail.com>
1 彭东林 <pengdonglin@smartisan.com>
1 彭浩(Richard) <richard.peng@oppo.com>
1 洪一竹 <sam.hung@emc.com.tw>
1 王振杰 <zhenjie.wang@sjtu.edu.cn>
1 王文虎 <wenhu.wang@vivo.com>
1 王程刚 <wangchenggang@vivo.com>
1 说不得 <gavin.kx@qq.com>
1 马建朋 <majianpeng@gmail.com>
1 黄乐 <huangle1@jd.com>
1 김규원 <chammoru@gmail.com>
1 남영민 <youngmin.nam@samsung.com>
1 배석진 <soukjin.bae@samsung.com>
1 이건호 <urpapa@gmail.com>
1 추지호 <jiho.chu@samsung.com>
```

</details>